aikido_perception/src/AprilTagsModule.cpp

```diff
@@ -0,0 +1,170 @@
+ */
+
+#include <aikido/perception/AprilTagsModule.hpp>
+#include <aikido/perception/shape_conversions.hpp>
+#include "dart/common/Console.h"
+#include <visualization_msgs/Marker.h>
+#include <visualization_msgs/MarkerArray.h>
+#include <ros/topic.h>
+#include <ros/ros.h>
+#include <ros/console.h>
+#include <boost/make_shared.hpp>
+#include <Eigen/Geometry>
+#include <tf/LinearMath/Vector3.h>
+#include <tf/LinearMath/Quaternion.h>
+#include <stdexcept>
+
+namespace aikido{
+namespace perception{
+
+//================================================================================================================================
+
+AprilTagsModule::AprilTagsModule( ros::NodeHandle node, std::string markerTopic, std::shared_ptr<ConfigDataLoader> configData,
+								  const dart::common::ResourceRetrieverPtr& resourceRetriever,	
+								  std::string referenceFrameId, dart::dynamics::Frame* referenceLink):
+		mNode(node),
+		mMarkerTopic(markerTopic),
+		mConfigData(configData),
+		mResourceRetriever(resourceRetriever),
+		mReferenceFrameId(referenceFrameId),
+		mReferenceLink(referenceLink)
+{
+	mListener = new tf::TransformListener();
+}
+
+//================================================================================================================================
+void AprilTagsModule::detectObjects(std::vector<dart::dynamics::SkeletonPtr>& skeleton_list,double _timeout, ros::Time timestamp)
+{
+	bool transform_lookedup = false;
+	//Looks at all detected tags, looks up config file 
+	//Appends new skeletons to skeleton list
+	do{
+		visualization_msgs::MarkerArrayConstPtr marker_message
+	 			= ros::topic::waitForMessage<visualization_msgs::MarkerArray>(mMarkerTopic,mNode,ros::Duration(_timeout));
+
+		for (const auto& marker_transform : marker_message->markers)
+		{
+			//Get marker, tag ID, and detection transform name 
+			const auto& marker_stamp = marker_transform.header.stamp;
+			const auto& tag_name = marker_transform.ns;
+			const auto& detection_frame = marker_transform.header.frame_id;
+
+			if(marker_stamp < timestamp){
+				dtwarn << "Marker stamp " << marker_stamp<< " for " << tag_name 
+				<< "is behind timestamp parameter: " <<timestamp <<"\n";
+
+				continue;
+			}
+
+			std::string skel_name;
+			Eigen::Isometry3d skel_offset;
+			dart::common::Uri skel_resource;
+
+			//check if tag name is in database
+			if (mConfigData->getTagNameOffset(tag_name,skel_name,skel_resource,skel_offset))
+			{
+
+				//Get orientation of marker
+				Eigen::Isometry3d marker_pose = aikido::perception::convertROSPoseToEigen(marker_transform.pose);
+
+
+				//For the frame-frame transform
+				tf::StampedTransform transform;
+
+				try{
+					mListener->waitForTransform(mReferenceFrameId,detection_frame,
+						marker_stamp,ros::Duration(_timeout));
+
+					mListener->lookupTransform(mReferenceFrameId,detection_frame,
+						marker_stamp, transform);
+				}
+				catch(const tf::ExtrapolationException& ex){
+					ROS_ERROR("%s",ex.what());
```

Comment:
I was not sure what you meant by this. I assume you meant a more detailed error message than `ROS_ERROR(ex.what())` so I've done something concatenating `ex.what()` to a context message and then sending its `.c_str()` to `ROS_ERROR`. Please have a look to confirm if that's what you meant.